### PR TITLE
Add more visibility to admins (part-2)

### LINKED ISSUE #19314: Add more visibility and control into the Admin Article page
## Is your feature request related to a problem? Please describe.
As a Community Admin, I need:
* more visibility into the inputs that impact an article’s score, and
* more control to make necessary moderation adjustments

## Describe the solution you'd like

Add a section to the Article Admin page found here `/admin/content_manager/articles/{article-id}`.

**Overview of Privileged actions with article title**:
* No. of Privileged actions at the top right
* Display the same overview in `/admin/content_manager/articles` as well
* **Score calculation**
  * The top line score represented by the article’s `article.score`
  * More granular visibility into the main inputs that go into that score…
  * The sum of reactions _on the author_, basically, taking the calculation from [this file](https://github.com/forem/forem/blob/main/app/models/article.rb#L559)
  * An aggregate sum point total of all “regular” reactions (❤️, 🦄, 🤯, 🙌, 🔥)
  * An aggregate sum point total of all “privileged” actions (👍, 👎, and 🧐)

Score hover

**Privileged Actions log**:
* A list of all 👍, 👎, and 🧐 reactions (privileged actions) - We will divide them into two tabs called `Flags` and `Quality reactions`. Note that 🧐 is called `vomit`
* **Flag tab**
  * Column representing the reaction category – reaction.category 🧐
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Columns representing the `status` of each privileged reaction. Will be `valid`, `invalid`, `Open`
    * Open -> Admin hasn't marked that action as valid/invalid
    * Valid -> Admin marked that action as valid
    * Invalid -> Admin marked that action as invalid
  * The ability to change the `status` of each action by the three-dot menu on the right.
  
  

* **Quality Reactions tab**
  * Column representing the reaction category – reaction.category 👍 👎
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Quality reactions will be the log of reactions only, and the admin won't be able to adjust them.
  
  
   
**For more details and mocks,** [**check Figma**](https://www.figma.com/file/yhg9HTZDFIeWuvBt9ygrss/Moderation?node-id=1485-4141&t=Pn3Q2r8QQoGd5Vqj-4).

## Requirements / Constraints

This can be a simplistic rendering with basic interactivity. We could add to this later, but the initial goal is to ensure more information and functionality is at least available to admins now, whereas previously it had been unavailable. We do not need to introduce any complicated client-side functionality.  Adjusting the `status` of reactions should take place with buttons that trigger a page reload.

## Additional context

This is the first step in a series of issues that will help our Community team.  Soon, we will also be adding similar functionality to:
* Comment admin page with similar functionality
* Adjust the functionality on the User’s “Flags” tab on the Users admin page

### PR DESCRIPTION
**Pull request by rt4914**
## What type of PR is this? (check all applicable)

- [x] Feature

## Description

Add a section to the Article Admin page found here `/admin/content_manager/articles/{article-id}`.

**Overview of Privileged actions with article title**:
* No. of Privileged actions at the top right
* Display the same overview in `/admin/content_manager/articles` as well
* **Score calculation**
  * The top line score represented by the article’s `article.score`
  * More granular visibility into the main inputs that go into that score…
  * The sum of reactions _on the author_, basically, taking the calculation from [this file](https://github.com/forem/forem/blob/main/app/models/article.rb#L559)
  * An aggregate sum point total of all “regular” reactions (❤️, 🦄, 🤯, 🙌, 🔥)
  * An aggregate sum point total of all “privileged” actions (👍, 👎, and 🧐)

**Privileged Actions log**:
* A list of all 👍, 👎, and 🧐 reactions (privileged actions) - We will divide them into two tabs called `Flags` and `Quality reactions`. Note that 🧐 is called `vomit`
* **Flag tab**
  * Column representing the reaction category – reaction.category 🧐
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Columns representing the `status` of each privileged reaction. Will be `valid`, `invalid`, `Open`
    * Open -> Admin hasn't marked that action as valid/invalid
    * Valid -> Admin marked that action as valid
    * Invalid -> Admin marked that action as invalid
  * The ability to change the `status` of each action by the three-dot menu on the right.

* **Quality Reactions tab**
  * Column representing the reaction category – reaction.category 👍 👎
  * Column representing the user who created that reaction – reactions.user_id
  * Column representing the point total of that individual reaction – reactions.points
  * Quality reactions will be the log of reactions only, and the admin won't be able to adjust them.

   
**For more details and mocks,** [**check Figma**](https://www.figma.com/file/yhg9HTZDFIeWuvBt9ygrss/Moderation?node-id=1485-4141&t=Pn3Q2r8QQoGd5Vqj-4).

## Related Tickets & Documents

- Closes #19314

## QA Instructions, Screenshots, Recordings

[**check Figma**](https://www.figma.com/file/yhg9HTZDFIeWuvBt9ygrss/Moderation?node-id=1485-4141&t=Pn3Q2r8QQoGd5Vqj-4).

All the details related to what this PR is, how to test it and what to expect has been covered in this video: 
[**How to test and what to expect?**](https://www.loom.com/share/4813e45d54104322a2f8bd7fa8b08624)

Screenshot 2023-05-04 at 5 57 00 PM

### Output
Screenshot 2023-05-02 at 5 33 19 PM
Screenshot 2023-05-02 at 5 33 28 PM
Screenshot 2023-05-02 at 5 33 34 PM
Screenshot 2023-05-02 at 5 33 48 PM
Screenshot 2023-05-02 at 5 34 11 PM
Screenshot 2023-05-02 at 5 34 19 PM
Screenshot 2023-05-04 at 3 07 53 PM

Screenshot 2023-05-02 at 5 42 03 PM
Screenshot 2023-05-02 at 5 45 35 PM
Screenshot 2023-05-02 at 5 45 41 PM
Screenshot 2023-05-02 at 5 48 27 PM

### UI accessibility concerns?

No

## Added/updated tests?

- [x] No

## [optional] Are there any post deployment tasks we need to perform?

@maestromac We should deploy this on canary1 or canary2 so that admin team can check the feature and provide suggestions.

### CODE REVIEW

**Comment by rt4914** [comment]
In this I am facing following error but unable to fix it correct.

```
GET /admin/content_manager/articles/1
| USE eager loading detected
|   Reaction => [:user]
|   Add to your query: .includes([:user])
| Call stack
|   …/forem/app/views/admin/articles/_flag_action_item.html
```

I tried solving this by adding following line:
`Bullet.add_safelist(type: :unused_eager_loading, class_name: "Reaction", association: :user)`
(Similar to https://github.com/forem/forem/pull/19274/files)

But this did not fix the issue.

**Comment by rt4914** [comment]
I have created this detailed video to show exactly what are the problems in my current implementation and what all things I have tried

https://www.loom.com/share/5a9f4e22012b458081b132ffd54e79fd

**Comment by rt4914** [comment]
Converted this PR to draft because there are some logical issues which needs to be fixed first.

**Comment by rt4914** [comment]
@maestromac Please deploy this on Canary1 or canary2 so that we can start testing it. Thanks.

File: app/views/admin/articles/_article_item.html.erb [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+     data-action="ajax:success@document->article#ajaxSuccess">
+
+  <header>
+    <% if decorated_article.pinned? || article.featured || article.approved || (!article.published? && article.published_from_feed?) || article.video || article.user&.warned? %>
+      <div class="flex gap-1 mb-1">
+        <% if !article.published? && article.published_from_feed? %>
+          <span class="c-indicator c-indicator--danger">Unpublished</span>
+        <% end %>
+        <% if decorated_article.pinned? %>
+          <span class="c-indicator c-indicator--warning" data-testid="pinned-indicator">Pinned</span>
```
Comment (Ridhwana) [inline app/views/admin/articles/_article_item.html.erb:18]: Do you think we should add these and some others to the i81n file?

Comment (rt4914) [reply app/views/admin/articles/_article_item.html.erb]: Applied this change. Shifted strings to yml files.

**Comment by rt4914** [comment]
@anujbhavsar96 I have removed the time as it was incorrect and also fixed the width or user name
Screenshot 2023-05-10 at 11 59 44 AM

**Comment by anujbhavsar96** [comment]
@rt4914, it would be great if we could display the date there. Of course, time is not that important, but the date can be helpful for admins to judge the overall impression of the article based on recent actions.

**Comment by rt4914** [comment]
> @rt4914, it would be great if we could display the date there. Of course, time is not that important, but the date can be helpful for admins to judge the overall impression of the article based on recent actions.

@anujbhavsar96 I have restored the date and time and made sure that when the status gets updated, the item will show the updated time too.
Screenshot 2023-05-10 at 2 14 30 PM

**Comment by Ridhwana** [comment]
@rt4914 just wanted to check if this is ready for review?

**Comment by Ridhwana** [comment]
> @Ridhwana From implementation perspective its finished. I am currently working on writing cypress tests for this.

Cool, could we move it into draft until those specs are done? 

> Also, I do have a query, I have pushed my code which contains a `qualityReactionArticle.spec.js`, basically as soon as I create a new article I want that there are some marked privileged reactions so that the vomits, high-quality and low-quality data is pre-filled. Any idea how can I do that? some sort of pre-created database?

If you need to add any seeds, you can do so in [the seeds_e2e](https://github.com/forem/forem/blob/b4a4b30153534ee7c9462d8417d74b4e33bf9e77/spec/support/seeds/seeds_e2e.rb) for Cypress/

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    cy.get('.crayons-dropdown').should('be.visible');
+  });
+
+  it('should open the dropdown on button click and close the dropdown on same button click', () => {
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('be.visible');
+
+    cy.get('.c-btn--icon-alone').click();
+    cy.get('.crayons-dropdown').should('not.visible');
+  });
```
Comment (maestromac) [inline cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:136]: E2E is way slower than unit test, so it's more acceptable to test more than one aspect per block. In this case, your `should open the dropdown on button click and close the dropdown on same button click` is already also testing what `should open the dropdown on button click` is doing, making the line 125 redundant.

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
```
Comment (maestromac) [inline cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:150]: Are these test duplicate of the above?

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  it('should display the open flag status', () => {
+    cy.get('.flex .c-indicator').should('be.visible').as('flagStatus');
+    cy.get('@flagStatus')
+      .should('have.class', 'c-indicator--warning')
+      .and('contain', 'Open');
+  });
+
+  it('should display the correct score for open flag', () => {
+    cy.get('.flex .crayons-card:nth-child(2) .fs-s').should('contain', '-50');
+  });
```
Comment (maestromac) [inline cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:123]: E2E is way slower than unit test, so it's more acceptable to test more than one aspect per block. In this case, these 4 specs can be merged into one and you can use comments to say what you are testing. ie
```
  it('should display the correct default data', () => {
    // show corrects user name
    ...

     // shows correct flag
     ...
  })
```

This way you check for all things in a single page load.

**Review by anujbhavsar96 — APPROVED** [review]
Everything works as expected 🚀. Great work @rt4914 ✨

**Comment by rt4914** [comment]
@Ridhwana @filleduchaos Looking for review on this PR. Thanks :-)

**Comment by Ridhwana** [comment]
@rt4914 there were some comments from Mac on the PR, are you still addressing those or can I go ahead and review?

**Comment by rt4914** [comment]
> @rt4914 there were some comments from Mac on the PR, are you still addressing those or can I go ahead and review?

@Ridhwana I completely missed those 3 comments in collapsed state. Just addressing them, overall they are not a blocker as such.

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js (hunk 2 above)
Comment (rt4914) [reply cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:216]: Updated the test to write it like this:
```
  it('should close the dropdown on same button click twice', () => {
    cy.get('.c-btn--icon-alone').click();
    cy.get('.c-btn--icon-alone').click();
    cy.get('.crayons-dropdown').should('not.visible');
  });
```

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js (hunk 3 above)
Comment (rt4914) [reply cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js:230]: They were partially duplicate which I have update now. But the main point is that in first test we are performing single click on button to open the dropdown and in second we are clicking it twice to make sure that the dropdown closes too.

File: cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js (hunk 4 above)
Comment (rt4914) [reply cypress/e2e/seededFlows/adminFlows/articles/privilegedReactions.spec.js]: Thanks for this suggestion. Based on this I have merged these tests to reduce them to almost half.

**Review by Ridhwana — APPROVED** [review]
LGTM 🚀

**Review by maestromac — APPROVED** [review]
Thanks for updating the test. LGTM!

**Pull request merged**